app/src/main/storage.ts

```diff
@@ -0,0 +1,112 @@
+import * as fs from "fs";
+import os from "os";
+import { ItemMetadata } from "../types";
+import { ItemFetchTask } from "./fetch/queue";
+
+/// Newtype for when we know a path component is potentially unsafe,
+/// which prevents it from being used in anything except PathBuilder
+export class UnsafePathComponent {
+  path: string;
+
+  constructor(path: string) {
+    this.path = path;
+  }
+}
+
+/**
+ * Class that helps build filesystem paths in a
+ * path-traversal safe way
+ */
+export class PathBuilder {
+  path: string;
+
+  constructor(path: string) {
+    if (!path.startsWith("/")) {
+      throw new Error("Root path must be absolute and start with /");
+    }
+    if (!path.endsWith("/")) {
+      throw new Error("Root path must be a directory and end with /");
+    }
+    this.path = path;
+  }
+
+  validate(part: string | UnsafePathComponent): string {
+    if (part instanceof UnsafePathComponent) {
+      part = part.path;
+    }
+    // Stop any path traversal attempts
+    if (part.includes("/") || part === ".." || part === "." || part === "") {
+      throw new Error(`Invalid path component: ${part}`);
+    }
+    return part;
+  }
+
+  doesntEscape(path: string) {
+    // Don't allow escape via symlinks
+    if (fs.existsSync(path)) {
+      const absolute = fs.realpathSync(path);
+      if (!absolute.startsWith(this.path)) {
+        throw new Error(`Path ${path} escapes root ${this.path}`);
+      }
+    }
+  }
+
+  getSubBuilder(...args: string[]): PathBuilder {
+    let ret = this.path;
+    for (const arg of args) {
+      this.validate(arg);
+      ret += `${arg}/`;
+    }
+    this.doesntEscape(ret);
+    return new PathBuilder(ret);
+  }
+
+  join(...args: (string | UnsafePathComponent)[]): string {
+    let ret = this.path;
+    for (let i = 0; i < args.length; i++) {
+      const validated = this.validate(args[i]);
+      ret += validated;
+      // Add separator between args (but not after the last one)
+      if (i < args.length - 1) {
+        ret += "/";
+      }
+    }
+    this.doesntEscape(ret);
+    return ret;
+  }
+}
+
+export class Storage {
+  private persistent: PathBuilder;
+  private tmp: PathBuilder;
+
+  constructor() {
+    this.persistent = new PathBuilder(os.homedir() + "/").getSubBuilder(
+      ".config",
+      "SecureDrop",
+      "files",
+    );
+    this.tmp = new PathBuilder(os.tmpdir() + "/");
```

Comment:
The temp path uses `os.tmpdir()` which is a directory in `/tmp` in Linux. It might be better to make the temp path `~/.config/SecureDrop/tmp/` and clean it up afterwards. In Qubes AppVMs, the private disk (`/rw`) contains the homedir, but `/tmp` is a tmpfs with I believe the default size of 1GB, so it risks running out of space.

Reply:
Hmm, in the client we already use /tmp for temporarily storing the downloads, and given that SD has a max cap server-side on 500MB files, I think we're probably okay, it would take multiple failed downloads to pile up before causing a problem. And I'd rather ensure we get the ephemeral properties of a real /tmp instead of needing to manage it ourselves.